tests/wazero/main.go

```diff
@@ -60,15 +57,15 @@ func main(){
 }
 
 type nonBlockReader struct {
-	buf *bytes.Buffer
+	buf    *bytes.Buffer
 	closed bool
 }
 
 func newNonBlockReader(r io.Reader) io.Reader {
```

Comment:
@ktock this is very curious.. can you explain more about why this is using a goroutine? it seems more about go than wazero, but I may be missing something.

Reply:
@codefromthecrypt Thanks for the question. We currently do non-blocking reading of stdin using `select` + `read` of wasi-sdk. When we run it on wazero, we observed that `select` reports a read event on stdin but `read`ing stdin blocks. So we currently forcibly make stdin non-blocking on the host side.
I'm not familiar to wazero internal so maybe I'm missing something. Suggestions are very welcome if you have better idea about implementing it.

Reply:
It seems that wasi-sdk [ignores `poll_oneoff` error other than EBADF](https://github.com/WebAssembly/wasi-libc/blob/wasi-sdk-19/libc-bottom-half/cloudlibc/src/libc/sys/select/pselect.c#L102) and wazero [returns non-EBADF error](https://github.com/tetratelabs/wazero/blob/v1.0.0-rc.2/imports/wasi_snapshot_preview1/poll.go#L135-L144) so I *guess* the error is ignored by wasi-sdk maybe?

Reply:
I think this is a nuanced area.. if you have %.wat, or a wasm binary ideally with source, that can reproduce the issue we can probably at least get to the bottom of it

Reply:
@codefromthecrypt  Thanks for the reply. This is the reproducer:

Sample C program:

```c
#include <stdio.h>
#include <unistd.h>
#include <sys/select.h>

int main(int argc, char **argv)
{
  char buf[5];
  int len = 5, ret;
  fd_set rfds;
  struct timeval tv;

  FD_ZERO(&rfds);
  FD_SET(0, &rfds);

  tv.tv_sec = 0;
  tv.tv_usec = 0;
  ret = select(1, &rfds, NULL, NULL, &tv);
  if ((ret > 0) && FD_ISSET(0, &rfds)) {
    printf("reading from stdin (select returned %d)\n", ret);
    fflush(stdout);
    ret = read(0, buf, len);
    if (ret > 0) {
      printf("got %s\n", buf);
      fflush(stdout);
    }
  } else {
    printf("no input\n");
  }
}
```

wasm file: [out.tar.gz](https://github.com/ktock/container2wasm/files/10998927/out.tar.gz)

wazero code:

```go
package main

import (
	"context"
	"os"

	"github.com/tetratelabs/wazero"
	"github.com/tetratelabs/wazero/imports/wasi_snapshot_preview1"
	"github.com/tetratelabs/wazero/sys"
)

func main() {
	ctx := context.Background()
	c, err := os.ReadFile(os.Args[1])
	if err != nil {
		panic(err)
	}
	r := wazero.NewRuntime(ctx)
	defer func() {
		r.Close(ctx)
	}()
	wasi_snapshot_preview1.MustInstantiate(ctx, r)
	compiled, err := r.CompileModule(ctx, c)
	if err != nil {
		panic(err)
	}
	_, err = r.InstantiateModule(ctx, compiled, wazero.NewModuleConfig().WithStdout(os.Stdout).WithStderr(os.Stderr).WithStdin(os.Stdin))
	if err != nil {
		if exitErr, ok := err.(*sys.ExitError); ok && exitErr.ExitCode() == 0 {
			return
		}
		panic(err)
	}
	return
}
```

wasm file is built using wasi-sdk 19.0

```
clang --sysroot=${WASI_SDK_PATH}/share/wasi-sysroot -o /out.wasm /main.c
```

On wazero, `select` reports read event even if nothing is sent to stdin.
`read` tries to read from stdin then blocks unexpectedly.

```
$ go run main.go ./out/out.wasm
reading from stdin (select returned 1)
```

It works properly on the host without block.

```
$ gcc main.c && ./a.out
no input
```

Reply:
@evacchi or @mathetake wanna take this on? there's still time to figure out an alternative before 1.0 next week. I'll be in crazy travel displacement mon,tues

Reply:
I'll try to take a look

Reply:
If I am following correctly, `select()` is returning 1 in wazero, while it should return 0; if I am reading `pselect()` right, this may only happen here:

```
  return readfds->__nfds + writefds->__nfds;
```

which means either is nonzero 🤔 :mono

Reply:
https://github.com/ktock/container2wasm/pull/23#discussion_r1139690107 might be related?

Reply:
I think the underlying issue is that we are only checking if the file descriptor exists (trivially true for stdin) instead of checking if there are bytes available

Reply:
WIP, but in the meantime you can try this https://github.com/tetratelabs/wazero/pull/1251